### PR TITLE
Adding OutIn Muon Step in Phase-2 tracking

### PR DESCRIPTION
**Pull request by ebrondol**
With this PR, the OutIn iteration has been added back into the tracking iteration sequence - which was cut out after the updating on the OT local reconstruction.
It has been tested in 910pre2 release and with single muon 0PU the gain is already pretty clear (blue - baseline; red - baseline+PR):
screenshot from 2017-05-03 13 50 41
Fakes have been tested using high PU events and changes are minor.

In principle, no change is expected in current tracking, but since `OutsideInMuonSeeder.cc` has been touched, it would be wise to double check.

### CODE REVIEW

**Comment by ebrondol** [comment]
@HuguesBrun please have a look

**Comment by dmitrijus** [comment]
+1

**Comment by ebrondol** [comment]
More validation plots can be found [here](https://cms-tracking-validation.web.cern.ch/cms-tracking-validation/PR/CMSSW_9_1_0_pre2_PR18553/).

**Comment by slava77** [comment]
@ebrondol 
please add some numbers for timing with PU200

Your MTV summary for PU200 seems to indicate a loss in efficiency. E.g. in HP:
``` Efficiency 	0.9038 -> 	0.8912```
More than 1% loss is not good. It's unclear how significant this is given 10 events.
It would be better to see the output is OK on a larger sample.

**Comment by VinInn** [comment]
if adding an iteration lower "efficiency", something fishy is happening downstream (including fake classification)
10 events are simply not enough (there are what <10 outin muons in total?) to conclude anything besides that no major havoc occur.

In any case the OutIn iteration is needed, no matter the cost (and the cost is surely negligible given the number of built tracks)

**Comment by ebrondol** [comment]
In fact, I have used the PU200 sample to check fakes and the PU0 sample to check efficiency.
You can notice the increment of efficiency in the single mu 0PU sample.
I have also update the page with ttbar 0PU sample, also in that case the increment of efficiency is there in HP:
Efficiency	0.9189 ->	0.9190
Number of matched TrackingParticles	464460 -> 464476

**Comment by slava77** [comment]
assign upgrade

I think that a check on a larger sample of PU200 is simple enough that it can be provided as a part of this PR.

**Comment by makortel** [comment]
I ran a test with 1k ttbar+35PU events in phase1 with this PR and added the plots to Erica's link
https://cms-tracking-validation.web.cern.ch/cms-tracking-validation/PR/CMSSW_9_1_0_pre2_PR18553/

Thins are mostly unchanged, only the fakes of muon outin show tiny changes
https://cms-tracking-validation.web.cern.ch/cms-tracking-validation/PR/CMSSW_9_1_0_pre2_PR18553/TTbar_PU35_Phase1_1kevents/plots_muonSeededStepOutIn/dupandfakePtEtaPhi.pdf

**Comment by slava77** [comment]
On 5/4/17 5:33 AM, Matti Kortelainen wrote:
> I ran a test with 1k ttbar+35PU events in phase1 with this PR and added
> the plots to Erica's link
> https://cms-tracking-validation.web.cern.ch/cms-tracking-validation/PR/CMSSW_9_1_0_pre2_PR18553/
> 
> Thins are mostly unchanged, only the fakes of muon outin show tiny changes
> https://cms-tracking-validation.web.cern.ch/cms-tracking-validation/PR/CMSSW_9_1_0_pre2_PR18553/TTbar_PU35_Phase1_1kevents/plots_muonSeededStepOutIn/dupandfakePtEtaPhi.pdf

This is the second derivative, if I'm not misunderstanding.
A comparison of differences in OutIn implementation between phase-1 and
phase-2 implementation applied to phase-1.
right?

**Comment by makortel** [comment]
> This is the second derivative, if I'm not misunderstanding.
> A comparison of differences in OutIn implementation between phase-1 and
> phase-2 implementation applied to phase-1.
> right?

My test shows what happens in phase1 with this PR. I guess the tiny change is caused by the changes in `OutsideInMuonSeeder.cc`.

**Comment by slava77** [comment]
On 5/4/17 5:47 AM, Matti Kortelainen wrote:
>     This is the second derivative, if I'm not misunderstanding.
>     A comparison of differences in OutIn implementation between phase-1 and
>     phase-2 implementation applied to phase-1.
>     right?
> 
> My test shows what happens in phase1 with this PR. I guess the tiny
> change is caused by the changes in |OutsideInMuonSeeder.cc|.

Ah, I didn't realize that this PR also changes behavior in phase-1/0.
Erica, please update the description or change the code to remove the
regression in phase-1 (even if it's tiny).
A bit more care will be needed with checkout if this PR stays changing
phase-1/0.

File: RecoTracker/SpecialSeedGenerators/src/OutsideInMuonSeeder.cc [hunk 1]
```diff
@@ -178,10 +184,19 @@ OutsideInMuonSeeder::produce(edm::Event & iEvent, const edm::EventSetup & iSetup
...
+            if (tmpTkGeometry->isThere(GeomDetEnumerators::P2OTEC)) {
+              LogDebug("OutsideInMuonSeeder") << "\n We are using the Phase2 Outer Tracker (defined as a TID). ";
+              forwLayers = (measurementTracker->geometricSearchTracker()->posTidLayers());
+            }
+            LogTrace("OutsideInMuonSeeder") << "\n ==== TEC tot layers " << forwLayers.size() << " ====" << std::endl;
+            int found = 0;
+            int iLayer = forwLayers.size(); 
+            if(iLayer==0) LogError("OutsideInMuonSeeder") << "TEC+ has no layers." ;
+
+            if (debug_) std::cout << "\n ==== Tot layers " << forwLayers.size() << " ====" << std::endl;
```
Comment (kpedro88) [inline RecoTracker/SpecialSeedGenerators/src/OutsideInMuonSeeder.cc]: Why is there a mix between `LogDebug` and cout statements behind `debug_` flags? It seems like `LogDebug` should be used everywhere.

File: RecoTracker/SpecialSeedGenerators/src/OutsideInMuonSeeder.cc [hunk 2]
```diff
@@ -178,10 +184,19 @@ OutsideInMuonSeeder::produce(edm::Event & iEvent, const edm::EventSetup & iSetup
             }
         }
         if (tk.eta() > minEtaForTEC_) {
-            int iLayer = 9, found = 0;
-            std::vector< ForwardDetLayer const* > const & tec = measurementTracker->geometricSearchTracker()->posTecLayers();
-            for (auto it = tec.rbegin(), ed = tec.rend(); it != ed; ++it, --iLayer) {
-                if (debug_) std::cout << "\n ==== Trying TEC " << +iLayer << " ====" << std::endl;
+            auto forwLayers = measurementTracker->geometricSearchTracker()->posTecLayers();
```
Comment (kpedro88) [inline RecoTracker/SpecialSeedGenerators/src/OutsideInMuonSeeder.cc]: Could this be `const auto&`?

File: RecoTracker/SpecialSeedGenerators/src/OutsideInMuonSeeder.cc [hunk 3]
```diff
@@ -191,10 +206,19 @@ OutsideInMuonSeeder::produce(edm::Event & iEvent, const edm::EventSetup & iSetup
             }
         }
         if (tk.eta() < -minEtaForTEC_) {
-            int iLayer = 9, found = 0;
-            std::vector< ForwardDetLayer const* > const & tec = measurementTracker->geometricSearchTracker()->negTecLayers();
-            for (auto it = tec.rbegin(), ed = tec.rend(); it != ed; ++it, --iLayer) {
-                if (debug_) std::cout << "\n ==== Trying TEC " << -iLayer << " ====" << std::endl;
+            auto forwLayers = measurementTracker->geometricSearchTracker()->negTecLayers();
```
Comment (kpedro88) [inline RecoTracker/SpecialSeedGenerators/src/OutsideInMuonSeeder.cc]: Could this be `const auto&`?

File: RecoTracker/TkDetLayers/src/Phase2OTtiltedBarrelLayer.cc [hunk 4]
```diff
@@ -97,13 +97,15 @@ Phase2OTtiltedBarrelLayer::groupedCompatibleDetsV( const TrajectoryStateOnSurfac
   vector<DetGroup> closestResultNeg;
   vector<DetGroup> closestResultPos;
   Phase2OTBarrelLayer::groupedCompatibleDetsV(tsos, prop, est, closestResultRods);
-  for(auto ring : theNegativeRingsComps){
-    ring->groupedCompatibleDetsV(tsos, prop, est, closestResultNeg);
-  }
-  for(auto ring : thePositiveRingsComps){
-    ring->groupedCompatibleDetsV(tsos, prop, est, closestResultPos);
+  if(tsos.globalPosition().z()<0){
+    for(auto ring : theNegativeRingsComps){
```
Comment (kpedro88) [inline RecoTracker/TkDetLayers/src/Phase2OTtiltedBarrelLayer.cc]: Should this be `auto&`? (also below)

File: RecoTracker/SpecialSeedGenerators/src/OutsideInMuonSeeder.cc (hunk 1 above)
Comment (ebrondol) [inline RecoTracker/SpecialSeedGenerators/src/OutsideInMuonSeeder.cc]: @kpedro88 the code is not under my responsability and I tried to keep the code as it was where was not necessary. @HuguesBrun can comment on that, I would be happy to clean up the `std::cout`.

File: RecoTracker/TkDetLayers/src/Phase2OTtiltedBarrelLayer.cc (hunk 4 above)
Comment (VinInn) [inline RecoTracker/TkDetLayers/src/Phase2OTtiltedBarrelLayer.cc]: is a pointer....

File: RecoTracker/SpecialSeedGenerators/src/OutsideInMuonSeeder.cc (hunk 2 above)
Comment (VinInn) [inline RecoTracker/SpecialSeedGenerators/src/OutsideInMuonSeeder.cc]: no, at most should be something like
auto forwLayers = tmpTkGeometry->isThere(GeomDetEnumerators::P2OTEC) ?
             measurementTracker->geometricSearchTracker()->posTidLayers()
           : measurementTracker->geometricSearchTracker()->posTecLayers();
(to be verified if const & compiles in such a case)

File: RecoTracker/SpecialSeedGenerators/src/OutsideInMuonSeeder.cc (hunk 3 above)
Comment (VinInn) [inline RecoTracker/SpecialSeedGenerators/src/OutsideInMuonSeeder.cc]: same as above

**Comment by VinInn** [comment]
Apparently OutIn is not reproducible even at Phase1 in 910.
I suggest to merge this PR as is.  In any case OutIn will require some cleanup that can only clash
with the modifications implemented in this PR.

**Comment by perrotta** [comment]
@VinInn What you say absolutely makes sense: I already examined the code changes in this PR, and indeed I couldn't find any of them possibly responsible of the lack of reproducibility. Therefore, I think this PR can be signed by reco. Just to collect here the complete documentation, could you please attach here the evidence that OutIn is not reproducible even at Phase1 in 910  (or a link to it)? Thx

File: RecoTracker/SpecialSeedGenerators/src/OutsideInMuonSeeder.cc (hunk 2 above)
Comment (kpedro88) [inline RecoTracker/SpecialSeedGenerators/src/OutsideInMuonSeeder.cc]: Yes, I discussed that separately with @ebrondol once she pointed out that the variable could take different values. I think `const auto&` works for this usage.

File: RecoTracker/SpecialSeedGenerators/src/OutsideInMuonSeeder.cc (hunk 1 above)
Comment (HuguesBrun) [inline RecoTracker/SpecialSeedGenerators/src/OutsideInMuonSeeder.cc]: Hi @ebrondol and @kpedro88 , the cout are here only for 'historical' reason so I see no problem to move to LogDebug. @ebrondol , it is great if you could clean up the cout !

**Comment by VinInn** [comment]
for Muon non-reproducibility please see issuse #18605

**Comment by VinInn** [comment]
@perrotta, well in OutsideInMuonSeeder.cc there is a prototypical possible source of non-reproducibility in multi-thread jobs...
I have proven that it is not the case and, when this PR is integrated, submit modifications that insure it cannot happen....

**Comment by perrotta** [comment]
Ok, Then the source of non-reproducibility stays outside the modifications submitted with this PR. Therefore, I think it can be accepted as such, now.

Before doing so, could @ebrondol please let us know which are the plans about
https://github.com/cms-sw/cmssw/pull/18553#discussion_r115178007 ? Do you plan to remove the cout's already with an update to this PR?

I would suggest to merge this PR, and prepare at once a new one which moves all cout's into a LogDebug. But please either update this one now or let us know.

**Comment by VinInn** [comment]
I suggest merge asap as I have mods already ready (and will take care of cout cleanup)

**Comment by ebrondol** [comment]
I am ready to push a commit with the comments/debug solved. I wanted to test it before pushing. Give me few minutes.

**Comment by ebrondol** [comment]
@kpedro88 please, let me know if now it is fine.
@HuguesBrun I changed `std::cout` but the variable `debug_` is still there. Please, take care of it in a second PR, to me it makes no sense to still have it.

**Comment by VinInn** [comment]
@cmsbuild, please test

**Comment by perrotta** [comment]
+1
OutIn iteration added to Phase2 tracking.
In https://github.com/cms-sw/cmssw/pull/18553#issuecomment-298925368 validation plots show the expected improvements on phase 2 muon tracking
No effect on phase 0/1 tracking is expected: the tiny regression observed is not caused by this PR and will be dealt with separately
The replacement of "cout" statements with LogDebug messages should be now finalized with the removal of the debug_ variable from config: this also will be included in a forthcoming PR

**Comment by slava77** [comment]
On 5/8/17 4:55 AM, perrotta wrote:
> In #18553 (comment)
> <https://github.com/cms-sw/cmssw/pull/18553#issuecomment-298925368>
> validation plots show the expected improvements on phase 2 muon tracking

the validation plots referenced here still show a drop in efficiency in
the sample with pileup.

**Comment by perrotta** [comment]
In PU 200 the efficiency is calculated over 230095 TrackingParticles for the reference, and 478 TrackingParticles for the release with this PR merged. We can't decide with it if there is a "drop in efficiency" or just a stat fluctuation:
- Number of TrackingParticles (after cuts)	230095	478
- Number of matched TrackingParticles	        210883	434

I think that such a comparison can only show that nothing fishy happens even with 200 PU (as also @VinInn said a few entries above), while a more detailed comparison is presented for the other samples

**Comment by slava77** [comment]
On 5/8/17 6:03 AM, perrotta wrote:
> In PU 200 the efficiency is calculated over 230095 TrackingParticles for
> the reference, and 478 TrackingParticles for the release with this PR
> merged. We can't decide with it if there is a "drop in efficiency" or
> just a stat fluctuation:
> 
>   * Number of TrackingParticles (after cuts) 230095 478
>   * Number of matched TrackingParticles 210883 434
> 
> I think that such a comparison can only show that nothing fishy happens
> even with 200 PU (as also @VinInn <https://github.com/VinInn> said a few
> entries above), while a more detailed comparison is presented for the
> other samples

I missed that the comparison was done on inconsistent samples.
Can we get a consistent comparison for self-documentation purpose?

**Comment by kpedro88** [comment]
+1

**Comment by ebrondol** [comment]
@slava77 
I have updated the validation website with 10 events also for the reference:
https://cms-tracking-validation.web.cern.ch/cms-tracking-validation/PR/CMSSW_9_1_0_pre2_PR18553/TTbar_PU200_10events/plots_ootb/effandfake1.pdf

**Comment by slava77** [comment]
On 5/8/17 7:48 AM, ebrondol wrote:
> @slava77 <https://github.com/slava77>
> I have updated the validation website with 10 events also for the reference:
> https://cms-tracking-validation.web.cern.ch/cms-tracking-validation/PR/CMSSW_9_1_0_pre2_PR18553/TTbar_PU200_10events/plots_ootb/effandfake1.pdf
> 

Thank you.
The summary now shows essentially no difference.
Good.

**Pull request merged**